doc/source/code_style.txt

```diff
@@ -0,0 +1,209 @@
+/******************************************************************************
+ *
+ * 
+ *
+ * Copyright (C) 2016
+ *
+ * Permission to use, copy, modify, and distribute this software and its
+ * documentation under the terms of the GNU General Public License is hereby 
+ * granted. No representations are made about the suitability of this software 
+ * for any purpose. It is provided "as is" without express or implied warranty.
+ * See the GNU General Public License for more details.
+ *
+ * Documents produced by Doxygen are derivative works derived from the
+ * input used in their production; they are not affected by this license.
+ *
+ */ /*! \page code_style Code Style for Contributors
+
+# Introduction #
+
+This document describes the code style to be used when contributing to
+the PIO library.
+
+Consistency of style allows bugs to be spotted more quickly and
+contributes to code quality.
+
+# C Code #
+
+## Backward Compatibility ##
+
+<p>We cannot remove or change a function in the public C API. We can
+add to it, carefully.
+
+## Brevity ##
+
+<p>Consider: C is a terse language.
+
+<ul>
+<li>Use the fact that NULL == 0, and 0 == false, to shorten code.
+<li>Assign and check return values in the same if statement.
+</ul>
+
+## Indentation and Spacing##
+
+<ul>
+<li>Use spaces not tabs.
+<li>4 spaces is the unit of intendation.
+<li>Indentation as defined by the "linux" style in emacs (see below).
+<li>Use spaces around most operators (=+-*/) not pointer or prefix/postfile (*++--)
+<li>Spaces after most keywords (if, for, while, etc.)
+<li>No spaces after function name.
+</ul>
+
+## Braces ##
+
+<p>Put braces on their own line, avoiding their use if possible.
+
+## Documentation ##
+
+<ul>
+<li>Every function must be documented using doxygen.
+<li>Use the doxygen \@internal to mark internal functions.
+<li>Use the doxygen \@ingroup to put public functions in the correct
+group.
+<li>Code must be reasonably documented as to intention.
+<li>Documentation quality and quantity are part of code review
+process.
+<li>Document in complete sentences.
+<li>Use C (not C++) comment delimiters.
+</ul>
+
+## Emacs ##
+
+<p>Put this in your .emacs file:
+
+<pre>
```

Comment:
What I have used on other projects, and would like to implement here, is a (voluntary) filter that is applied when code is checked in and applies formatting as automatically as possible.

In some environments (like eclipse) this is trivially easy to set up, but in emacs it will require a little messing about. But the idea is that everything that can be done automatically is done whenever a file is committed, and so the code in the repo is always meeting the standard.

So if you want to come up with an emacs function that runs untabify, and indents-region, that would be cool and we can add it to the document.